File: finance_app/lib/utils/message_parser.dart
import 'dart:math';

class ParsedTransaction {
  final double? amount;
  final String? merchant;
  final String? categoryHint;
  final DateTime? date;

  ParsedTransaction({
    this.amount,
    this.merchant,
    this.categoryHint,
    this.date,
  });

  bool get isValid => amount != null && amount! > 0;
}

class MessageParser {
  // Regex patterns for common Indian SMS formats
  static final RegExp _amountRegex =
      RegExp(r'(?:₹|Rs\.?|INR)\s*([\d,]+\.?\d*)', caseSensitive: false);

  static final RegExp _debitRegex =
      RegExp(r'(?:debited|paid|charged|transferred)', caseSensitive: false);

  static final RegExp _creditRegex =
      RegExp(r'(?:credited|received|refunded)', caseSensitive: false);

  static final RegExp _merchantRegex =
      RegExp(r'(?:at|to|from)\s+([A-Za-z\s]+?)(?:via|on|through|by|\.|$)', caseSensitive: false);

  // Common category hints
  static const Map<String, String> _categoryMappings = {
    'amazon': 'Shopping',
    'flipkart': 'Shopping',
    'swiggy': 'Food',
    'zomato': 'Food',
    'uber': 'Transport',
    'ola': 'Transport',
    'irctc': 'Travel',
    'makemytrip': 'Travel',
    'netflix': 'Entertainment',
    'hotstar': 'Entertainment',
    'electricity': 'Bills',
    'water': 'Bills',
    'gas': 'Bills',
    'phone': 'Bills',
    'mobile': 'Bills',
    'airtel': 'Bills',
    'vodafone': 'Bills',
    'jio': 'Bills',
  };

  ParsedTransaction parse(String message) {
    message = message.toLowerCase();

    double? amount = _extractAmount(message);
    String? merchant = _extractMerchant(message);
    String? categoryHint = merchant != null ? _guessCategory(merchant) : null;
    String transactionType = _debitRegex.hasMatch(message) ? 'Expense' :
                           _creditRegex.hasMatch(message) ? 'Income' : 'Expense'; // Default to expense

    // For now, assume date is today; in future could parse date from message
    DateTime? date = DateTime.now();

    // Override categoryHint based on transaction type; if not expense, clear category
    if (transactionType != 'Expense') {
      categoryHint = null;
    }

    return ParsedTransaction(
      amount: amount,
      merchant: merchant,
      categoryHint: categoryHint,
      date: date,
    );
  }

  double? _extractAmount(String message) {
    final match = _amountRegex.firstMatch(message);
    if (match != null) {
      String amountStr = match.group(1)?.replaceAll(',', '') ?? '';
      return double.tryParse(amountStr);
    }
    return null;
  }

  String? _extractMerchant(String message) {
    final match = _merchantRegex.firstMatch(message);
    if (match != null) {
      String merchant = match.group(1)?.trim() ?? '';
      // Clean up common prefixes
      merchant = merchant.replaceAll(RegExp(r'^(the|a|an)\s+', caseSensitive: false), '');
      return merchant.isNotEmpty ? merchant : null;
    }
    return null;
  }

  String? _guessCategory(String merchant) {
    String cleanMerchant = merchant.toLowerCase();
    for (var entry in _categoryMappings.entries) {
      if (cleanMerchant.contains(entry.key)) {
        return entry.value;
      }
    }
    return null;
  }
}

File: finance_app_kotlin/app/src/main/java/com/tracker/finance_app/core/util/MessageParser.kt
package com.tracker.finance_app.core.util

data class ParsedTransaction(
    val amount: Double?,
    val merchant: String?,
    val categoryHint: String?,
    val transactionType: String = "Expense"
) {
    val isValid: Boolean get() = amount != null && amount > 0
}

object MessageParser {
    private val amountRegex = Regex("""(?:₹|Rs\.?|INR)\s*([\d,]+\.?\d*)""", RegexOption.IGNORE_CASE)
    private val debitRegex = Regex("""(?:debited|paid|charged|transferred)""", RegexOption.IGNORE_CASE)
    private val creditRegex = Regex("""(?:credited|received|refunded)""", RegexOption.IGNORE_CASE)
    private val merchantRegex = Regex("""(?:at|to|from)\s+([A-Za-z\s]+?)(?:via|on|through|by|\.|$)""", RegexOption.IGNORE_CASE)

    private val categoryMappings = mapOf(
        "amazon" to "Shopping",
        "flipkart" to "Shopping",
        "swiggy" to "Food",
        "zomato" to "Food",
        "uber" to "Transport",
        "ola" to "Transport",
        "irctc" to "Travel",
        "makemytrip" to "Travel",
        "netflix" to "Entertainment",
        "hotstar" to "Entertainment",
        "electricity" to "Bills",
        "water" to "Bills",
        "gas" to "Bills",
        "phone" to "Bills",
        "mobile" to "Bills",
        "airtel" to "Bills",
        "vodafone" to "Bills",
        "jio" to "Bills"
    )

    fun parse(message: String): ParsedTransaction {
        val lowerMessage = message.lowercase()
        val amount = extractAmount(lowerMessage)
        val merchant = extractMerchant(lowerMessage)
        val categoryHint = merchant?.let { guessCategory(it) }
        val type = when {
            debitRegex.containsMatchIn(lowerMessage) -> "Expense"
            creditRegex.containsMatchIn(lowerMessage) -> "Income"
            else -> "Expense"
        }
        return ParsedTransaction(amount, merchant, if (type == "Expense") categoryHint else null, type)
    }

    private fun extractAmount(message: String): Double? {
        return amountRegex.find(message)?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull()
    }

    private fun extractMerchant(message: String): String? {
        val match = merchantRegex.find(message) ?: return null
        val merchant = match.groupValues[1].trim()
            .replace(Regex("""^(the|a|an)\s+""", RegexOption.IGNORE_CASE), "")
        return merchant.ifBlank { null }
    }

    private fun guessCategory(merchant: String): String? {
        val clean = merchant.lowercase()
        return categoryMappings.entries.firstOrNull { clean.contains(it.key) }?.value
    }
}
